server/src/main/java/org/elasticsearch/action/admin/indices/rollover/AutoShardCondition.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.admin.indices.rollover;
+
+import org.elasticsearch.TransportVersion;
+import org.elasticsearch.TransportVersions;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+
+/**
+ * Condition for automatically increasing the number of shards for a data stream. This indicates the optimum number of shards that was
+ * configured for the index abstraction as part of rollover.
+ * It's more of a marker condition, when present the condition is met, more than a condition we evaluate against stats.
+ */
+public class AutoShardCondition extends Condition<Integer> {
+    public static final String NAME = "auto_sharding";
```

Comment:
I think now that this is only the optimal shard count, I think we should call it the `OptimalShardCountCondition` (and change the `NAME` to something like `optimal_shard_count`)